### PR TITLE
Add asset sell support to the RFQ service

### PR DESCRIPTION
**Pull request by ffranr**
Closes https://github.com/lightninglabs/taproot-assets/issues/818

This PR builds on top of the refactoring work from https://github.com/lightninglabs/taproot-assets/pull/814 .

It adds asset sell request and sell request accept support to the RFQ service.

### CODE REVIEW

**Comment by dstadulis** [comment]
There's local code
need to rebase code
getting the itest

**Comment by ffranr** [comment]
I'm not sure how to setup the itest for the case where we sell a tap asset as the payment to satisfy the first hop in a lightning payment route. 

Alice wants to pay Carol. Carol generates the invoice.

Alice wants to sell a tap asset to Bob, and Alice would like Bob to send the bid payment in that sale to Carol. In so doing Bob would satisfy the invoice that Carol generated for Alice.

I can currently:
* Reach a sale quote agreement between Alice and Bob.
* Generate an invoice with Carol.
* Generate a custom route from Alice, to Bob, and then to Carol.

Bob can use his RFQ service to listen for a HTLC from Alice that satisfies the agreed Alice sale quote. And Alice can create a custom route with an input amount of 0 to Bob, but then stipulate that Bob must pay Carol the balance agreed upon in Alice-Bob's sale quote.

Is that the way forward? @guggero ?

**Comment by guggero** [comment]
There are important parts missing on the `lnd` side for this to work fully. So we need to find a compromise so we can test something. I think what we discussed offline at some point was that for now Alice would embed custom data to Bob (the same way you can embed `dest_custom_records` for the final receiver, we'd want to include a custom TLV record for Bob, I think we found that the `SendToRoute` might allow that or might need some additions to allow for it).
The custom data at this moment could just be the RFQ ID. Then Bob's `tapd` could intercept the HTLC, decode the custom record data and check that an offer with that RFQ ID exists, then check that the outgoing amount matches that offer.

**Comment by ffranr** [comment]
> There are important parts missing on the `lnd` side for this to work fully. So we need to find a compromise so we can test something. I think what we discussed offline at some point was that for now Alice would embed custom data to Bob (the same way you can embed `dest_custom_records` for the final receiver, we'd want to include a custom TLV record for Bob, I think we found that the `SendToRoute` might allow that or might need some additions to allow for it). The custom data at this moment could just be the RFQ ID. Then Bob's `tapd` could intercept the HTLC, decode the custom record data and check that an offer with that RFQ ID exists, then check that the outgoing amount matches that offer.

@guggero In the asset buy payment initiation case the HTLC intercepting node uses the SCID to identify the relevant HTLC. Both peers are aware of the SCID as it is present in the agreed upon quote (derived from message ID). Why do we need custom records? Can't we just send forward a HTLC over the quote SCID?

**Comment by guggero** [comment]
> @guggero In the asset buy payment initiation case the HTLC intercepting node uses the SCID to identify the relevant HTLC. Both peers are aware of the SCID as it is present in the accepted quote (SCID is derived from message ID). Why do we need custom records? Can't we just forward a HTLC over the accepted quote SCID?

Oh, you're right, we have the `incoming_channel_id`, I wasn't aware of that. I thought we'd only have the outgoing one, which in this case would be a "real" channel in the network. But then yes, we can just use the quote SCID to match things, no need for custom records.

**Comment by ffranr** [comment]
> > @guggero In the asset buy payment initiation case the HTLC intercepting node uses the SCID to identify the relevant HTLC. Both peers are aware of the SCID as it is present in the accepted quote (SCID is derived from message ID). Why do we need custom records? Can't we just forward a HTLC over the accepted quote SCID?
> 
> Oh, you're right, we have the `incoming_channel_id`, I wasn't aware of that. I thought we'd only have the outgoing one, which in this case would be a "real" channel in the network. But then yes, we can just use the quote SCID to match things, no need for custom records.

I've managed to get the custom records method to work with a custom route. But I didn't manage to forward a HTLC when I set `ChanId` on the custom route or set the incoming channel on a hop hint. Not sure what's going on there, perhaps LND is rejecting the channel ID given that an actual channel exists.

**Comment by ffranr** [comment]
In the last commit in this PR i bump lndclient so that I can inspect custom records with HTLC intercept. I'll spin out the bump commit into its own PR.

The first 9 commits can be split out into a separate PR. They are effectively refactor commits.

**Comment by dstadulis** [comment]
itests are failing when running together

**Comment by dstadulis** [comment]
@ffranr will split out PRs to get merge faster

**Comment by ffranr** [comment]
please review https://github.com/lightninglabs/taproot-assets/pull/845 first

**Comment by ffranr** [comment]
In today's standup, I talked about trying to fix a bug where two RFQ tests can't run at the same time. Turns out, the problem is with the lnd/lntest node manager – it just can't handle more than 7 nodes at once. If you try to add LND node numer 8, it'll just freeze up without giving you an error. This is probably because the default setting for the btcd miner RPC can only take 10 clients max, and adding another LND node would make it 11, which is one too many.

The simplest workaround for this issue, which likely necessitates modifying `lnd/lntest`, is to run the RFQ integration tests asynchronously. This approach allows us to progress to version 0.4 while planning to enhance `lntest` later. What are your thoughts on this strategy? Does anyone have deeper insights into `lntest`?

**Comment by guggero** [comment]
To bypass the issue in `lnd/lntest`, did you try if re-using the already running standby nodes (Alice and Bob) would help? Then you need to start two fewer nodes. There are `tapd` nodes connected to those, but they should behave pretty independently when it comes to channels and on-chain funds.

**Comment by ffranr** [comment]
> To bypass the issue in `lnd/lntest`, did you try if re-using the already running standby nodes (Alice and Bob) would help? Then you need to start two fewer nodes. There are `tapd` nodes connected to those, but they should behave pretty independently when it comes to channels and on-chain funds.

I will try what you've suggested. But I think just running the test asynchronously would be a simpler temporary solution. The RFQ tests create three new LND nodes and ensures they have the correct channel setup between them. Using one of the standby nodes may lead to unintended future side-effects from channel states or from HTLC interception.

**Comment by ffranr** [comment]
> To bypass the issue in `lnd/lntest`, did you try if re-using the already running standby nodes (Alice and Bob) would help? Then you need to start two fewer nodes. There are `tapd` nodes connected to those, but they should behave pretty independently when it comes to channels and on-chain funds.

I've just tried using standby node Alice as the third node in the RFQ tests. When running one of the RFQ itests in conjunction with the rest of the itests, a LND peer message channel (which the RFQ stream handler is subscribed to) closed unexpectedly.

I'm going to try just running one of the RFQ tests asynchronously to the other itests.

**Comment by ffranr** [comment]
Turns out that I wasn't cleaning up the RFQ test scenario LND nodes correctly. They were living on in the test harness LND node manager. I've corrected that and now the itests pass.

I still think that we should address the `lnd/lntest` node count limitation at one point, but that doesn't matter to tap too much right now. I don't think we'll need >=6 LND nodes simultaneously in a single itest anytime soon.

File: taprpc/rfqrpc/rfq.proto
Comment (guggero) [inline taprpc/rfqrpc/rfq.proto:91]: nit: "Unix timestamp in seconds"?

File: rfq/manager.go
Comment (guggero) [inline rfq/manager.go:301]: What additional information do we need to implement this TODO? Any reason not to do it in this PR?

File: taprpc/rfqrpc/rfq.proto
Comment (guggero) [inline taprpc/rfqrpc/rfq.proto:180]: nit: "in microseconds"!

File: taprpc/rfqrpc/rfq.proto
Comment (guggero) [inline taprpc/rfqrpc/rfq.proto:155]: nit: "in seconds".

File: rpcserver.go
Comment (guggero) [inline rpcserver.go:5726]: Don't we use microseconds in all our events?

File: itest/rfq_test.go
Comment (guggero) [inline itest/rfq_test.go:273]: I don't think that `wait.NoError()` is the right method here, since that will just re-try (without canceling anything on timeout), but `Recv()` will block until a message arrives.

File: itest/rfq_test.go
Comment (guggero) [inline itest/rfq_test.go:345]: Same here. I think a goroutine that sends on a channel, followed by a `select` on the channel with a timeout is probably what we want.

File: itest/rfq_test.go
Comment (guggero) [inline itest/rfq_test.go:365]: Can we implement this TODO in this PR? Or anything still missing?

**Review by guggero — COMMENTED** [review]
Looks pretty good, very nice work :tada: 
Just a few minor comments

File: rfq/negotiator.go
Comment (ffranr) [reply rfq/negotiator.go:412]: There's a TODO in this function that will end up with an error return code path:
```
// TODO(ffranr): Ensure that we have a suitable buy offer for the asset
//  that our peer is trying to sell to us.
```

It will end up looking like `HandleIncomingBuyRequest`.

File: itest/rfq_test.go
Comment (ffranr) [reply itest/rfq_test.go:345]: Can't we just block on the `bobEventNtfns.Recv()` call and depend on the itest timeout rather than using `select`? We expect to receive the event.

File: rfq/manager.go
Comment (ffranr) [reply rfq/manager.go:301]: In this PR I'm just trying to add functionality for the asset sell payment initiation to match what we already have for the asset buy payment initiation. The functionality that this TODO refers to is also missing for the incoming `rfqmsg.BuyAccept` case.

I can add this functionality in a small future PR.

File: rfq/negotiator.go
Comment (ffranr) [reply rfq/negotiator.go:361]: There's a TODO in this function that will end up with an error return code path:
```
// TODO(ffranr): Ensure that we have a suitable buy offer for the asset
//  that our peer is trying to sell to us.
```

It will end up looking like `HandleIncomingBuyRequest`.

File: itest/rfq_test.go
Comment (GeorgeTsagk) [inline itest/rfq_test.go:284]: What are Bob's acceptance criteria here?

File: rfqmsg/sell_accept.go
Comment (ffranr) [reply rfqmsg/sell_accept.go:70]: Going with just `records()`.

File: itest/rfq_test.go
Comment (GeorgeTsagk) [reply itest/rfq_test.go:365]: Could just assert the invoice here?

**Review by GeorgeTsagk — COMMENTED** [review]
Looks good, a few Qs
very nice commit structure

File: rfq/manager.go
Comment (ffranr) [reply rfq/manager.go:499]: The idea here is to (in the future) support submitting a sell order to the RFQ (Request for Quote) service without designating a specific peer. Consequently, the RFQ service would be responsible for identifying the most suitable peer, potentially by soliciting quotes from various peers to ascertain the most favorable offer.

File: rfq/negotiator.go
Comment (ffranr) [reply rfq/negotiator.go:306]: Going forward, we might send the ask price over to a price oracle so it can take that into account when coming up with a bid price. But since we haven't got a price oracle up and running yet, we're holding off on passing the ask price along for now.

File: rfq/negotiator.go
Comment (ffranr) [reply rfq/negotiator.go:372]: `queryAskFromPriceOracle` is supposed to be the function for the oracle-asking logic. The rest of the code creates a new request and puts it on the outgoing channel.

This is where I think something like this might come in handy:
```golang
func execGoRoutine(cg *fn.ContextGuard, errChan chan<- error,
    exec func() error) error {

    // Ensure that the error channel has a buffer of at least 1.
    if cap(errChan) == 0 {
        return fmt.Errorf("error channel has no buffer")
    }

    cg.Wg.Add(1)
    go func() {
        defer cg.Wg.Done()

        err := exec()
        if err != nil {
            sendSuccess := fn.SendOrQuit(errChan, err, cg.Quit)
            if !sendSuccess {
                log.Warn("Failed to send error on error "+
                    "channel: %v", err)
            }
        }
    }()
    
    return nil
}
```

I'll see if i can refactor later. Is this nit blocking for you?

File: itest/rfq_test.go
Comment (ffranr) [reply itest/rfq_test.go:284]: We don't have an acceptance criteria for Bob right now. Bob will just accept any offer.

There's a TODO in the `Negotiator` which will give Bob a criteria:
```
	// TODO(ffranr): Ensure that we have a suitable buy offer for the asset
	//  that our peer is trying to sell to us.
```

File: itest/rfq_test.go
Comment (ffranr) [reply itest/rfq_test.go:365]: hmm I can't get this to work for some reason. Route is: `Alice -> Bob -> Carol`, the HTLC does seem to reach Bob. And Bob does accept the intercepted HTLC. But then the send function returns:
```
Failure_INCORRECT_OR_UNKNOWN_PAYMENT_DETAILS (1)
```

And so when i lookup the invoice from Carol, it isn't settled.

Comment (ffranr) [reply itest/rfq_test.go:352]: I've attempted to disable Bob's tap HTLC interception but the payment still doesn't succeed. Any ideas @guggero @GeorgeTsagk ?

Comment (ffranr) [reply itest/rfq_test.go:352]: I've got the payment to succeed.

File: rfq/negotiator.go
Comment (guggero) [reply rfq/negotiator.go:361]: Makes sense :+1:

File: rfq/manager.go
Comment (guggero) [reply rfq/manager.go:301]: Sounds good to me.

File: rfqmsg/sell_request.go
Comment (ffranr) [reply rfqmsg/sell_request.go:134]: I've added basic encode/decode unit tests for the new message types.

**Comment by ffranr** [comment]
Some of the review comments also apply to existing code that the PR doesn't touch. I've made a note of them and I'll create another PR address those code changes.

File: itest/rfq_test.go
Comment (guggero) [reply itest/rfq_test.go]: Yes, but then we'd need to initiate `ts.AliceTapd.SubscribeRfqEventNtfns` with a timeout context. Otherwise it will just block forever, since there's nothing to time out (other than eventually `go test` timing out, which currently is set at 20 minutes with `-test.timeout`).

But I think the `BeforeTimeout` function works too!

**Review by guggero — APPROVED** [review]
Great work! LGTM pending unit tests for encode/decode of messages.

File: rfq/negotiator.go
Comment (GeorgeTsagk) [reply rfq/negotiator.go:372]: nop, looks good!

**Review by GeorgeTsagk — APPROVED** [review]
Looks good!

are the encode/decode unit tests landing in a diff PR?
otherwise lgtm

**Comment by ffranr** [comment]
@GeorgeTsagk thanks for taking a second look.

I've added these unit tests for encode/decode:
https://github.com/lightninglabs/taproot-assets/pull/835/files#diff-26f6b4e4b5386817099becddf822524f93feec1169c2f0901380dabb7c25d740

https://github.com/lightninglabs/taproot-assets/pull/835/files#diff-4ee5dc5905f2ad2d13635df68e16d1881e3ceedde4ab1f91529d7b95f0c7396a

**Pull request merged**
